src/main/java/com/google/growpod/servlets/GardenServlet.java

```diff
@@ -0,0 +1,162 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.growpod.servlets;
+
+import com.google.growpod.data.Garden;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that handles garden entities on the server.
+ *
+ * <p>API DOCUMENTATION: /garden/{id} {id} -- A garden UUID GET: Retrieves the garden data structure
+ * for {id} No parameters Returns data in JSON format along with (200 OK), otherwise (404 NOT FOUND)
+ * /garden/{id}/plant-list {id} -- A garden UUID GET: Retrieves all plants currently on {id} garden
+ * No parameters Returns data in JSON format along with (200 OK), otherwise (404 NOT FOUND)
+ * /garden/{id}/user-list {id} -- A garden UUID GET: Retrieves all users currently on {id} garden No
+ * parameters Returns data in JSON format along with (200 OK), otherwise (404 NOT FOUND)
+ */
+@WebServlet({"/garden", "/garden/*"})
```

Comment:
Do you know if it is possible to split this into multiple servlets so that you can have the framework handle more of the routing for you?

E.g.:
1. `"/garden"`
2. `"/garden/{id}"`
2. `"garden/{id}/user-list"`
2. `"garden/{id}/plant-list"`

etc.

Reply:
Probably not with this servlet API. The documentation on WebServlet's URL patterns is nonexistent though, so I don't know what would be allowed. https://javaee.github.io/javaee-spec/javadocs/javax/servlet/annotation/WebServlet.html#urlPatterns--

Reply:
I'd be very surprised if it isn't possible, because otherwise how could you compose multiple endpoints?

I think you need to make _multiple_ servlets (each with its own file) instead of just one.

E.g.
1. `GardenServlet`  - handles just `/garden`
2. `GardenIdServlet` - handles just `/garden/{id}`
...

Reply:
I tried to make a servlet that matches "/plant/*/id" (just for testing purposes) but I got the below message.
```
[INFO] GCLOUD: java.lang.IllegalArgumentException: Servlet Spec 12.2 violation: glob '*' can only exist at end of prefix based matches: bad spec "/plant/*/id"
```

There's another Java servlet API, "JAX-RS", that could handle the REST-ish API better.

I'll have to look how I would solve this problem under the constraints of this current API.

Reply:
Ok -- the Servlet Spec 12.2 in the error message, in this document https://javaee.github.io/servlet-spec/downloads/servlet-4.0/servlet-4_0_FINAL.pdf , explicitly does not allow for any wildcards in the middle of a mapping.

Reply:
Nice job finding the spec! My mind is blown that all of their examples are actually using plant/gardening examples. I.e. 
<img width="513" alt="Screen Shot 2020-07-07 at 5 49 58 PM" src="https://user-images.githubusercontent.com/4656974/86847782-e5abb500-c07a-11ea-8dc3-9d8200bbe8e9.png">.

Saddening that Servlets don't support this out of the box, but thank you for researching / getting to the bottom of it.